R/sim_linelist.R

```diff
@@ -0,0 +1,266 @@
+#'
+#' @examples
+#' # load data required to simulate line list
+#' contact_distribution <- epiparameter::epidist(
+#'   disease = "COVID-19",
+#'   epi_dist = "contact distribution",
+#'   prob_distribution = "pois",
+#'   prob_distribution_params = c(mean = 2)
+#' )
+#'
+#' infect_period <- epiparameter::epidist(
+#'   disease = "COVID-19",
+#'   epi_dist = "infectious period",
+#'   prob_distribution = "gamma",
+#'   prob_distribution_params = c(shape = 1, scale = 1)
+#' )
+#'
+#' # get onset to hospital admission from {epiparameter} database
+#' onset_to_hosp <- epiparameter::epidist_db(
+#'   disease = "COVID-19",
+#'   epi_dist = "onset to hospitalisation",
+#'   single_epidist = TRUE
+#' )
+#'
+#' # get onset to death from {epiparameter} database
+#' onset_to_death <- epiparameter::epidist_db(
+#'   disease = "COVID-19",
+#'   epi_dist = "onset to death",
+#'   single_epidist = TRUE
+#' )
+#' # example with single hospitalisation risk for entire population
+#' linelist <- sim_linelist(
+#'   contact_distribution = contact_distribution,
+#'   infect_period = infect_period,
+#'   prob_infect = 0.5,
+#'   onset_to_hosp = onset_to_hosp,
+#'   onset_to_death = onset_to_death,
+#'   hosp_risk = 0.5
+#' )
+#' head(linelist)
+#'
+#' # example with age-stratified hospitalisation risk
+#' # 20% for over 80s
+#' # 10% for under 5s
+#' # 5% for the rest
+#' age_dep_hosp_risk <- data.frame(
+#'   age_limit = c(1, 5, 80),
+#'   risk = c(0.1, 0.05, 0.2)
+#' )
+#' linelist <- sim_linelist(
+#'   contact_distribution = contact_distribution,
+#'   infect_period = infect_period,
+#'   prob_infect = 0.5,
+#'   onset_to_hosp = onset_to_hosp,
+#'   onset_to_death = onset_to_death,
+#'   hosp_risk = age_dep_hosp_risk
+#' )
+#' head(linelist)
+sim_linelist <- function(contact_distribution,
+                         infect_period,
+                         prob_infect,
+                         onset_to_hosp,
+                         onset_to_death,
+                         onset_to_recovery = NA,
+                         hosp_risk = 0.2,
+                         hosp_death_risk = 0.5,
+                         non_hosp_death_risk = 0.05,
+                         outbreak_start_date = as.Date("2023-01-01"),
+                         anonymise = FALSE,
+                         outbreak_size = c(10, 1e4),
+                         population_age = c(1, 90),
+                         case_type_probs = c(
+                           suspected = 0.2,
+                           probable = 0.3,
+                           confirmed = 0.5
+                         ),
+                         config = create_config()) {
+  # check and convert distribution to func if needed before .check_sim_input()
+  funcs <- as_function(
+    list(
+      contact_distribution = contact_distribution,
+      infect_period = infect_period,
+      onset_to_hosp = onset_to_hosp,
+      onset_to_death = onset_to_death,
+      onset_to_recovery = onset_to_recovery
+    )
+  )
```

Comment:
> [!NOTE]
> No action needed, just sharing thinking

I wish we could destructure lists more easily in R and have statements like this

```r
{ contact_distribution,
  ...,
  onset_to_recovery } <- as_function(list(contact_distribution,
    ...,
    onset_to_recovery))
```

It would save you the reassignment hassle in the lines afterward.

But I guess I got too used to JavaScript where this is possible 😅 https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

Reply:
Good point. There have been some attempts to do this in R (e.g. [{zeallot}](https://cran.r-project.org/web/packages/zeallot/index.html)), but I'll stick to long-winded unpacking of lists line-by-line.

> But I guess I got too used to JavaScript where this is possible 😅 https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment

(Always nice to learn how things are done in our languages to see if we can improve our R code).